### PR TITLE
Dev Workflow Using `Makefile` 

### PR DESCRIPTION
**Pull request by glenn-sorrentino**
- Add an option in the installer for the dev vs. prod Redis setting.
- move all environment variables to the `Makefile`
- Added `.pre-commit-config.yaml`
- Formatting cleanup

### CODE REVIEW

**Comment by glenn-sorrentino** [comment]
Closes #315

**Comment by glenn-sorrentino** [comment]
@brassy-endomorph Ready for review

**Comment by glenn-sorrentino** [comment]
Tested working executing `make run`:

Screenshot 2024-04-02 at 1 17 43 PM

Screenshot 2024-04-02 at 1 24 10 PM

File: Makefile [hunk 1]
```diff
@@ -1,32 +1,58 @@
...
+export FLASK_APP REDIS_URI SECRET_KEY ENCRYPTION_KEY DB_NAME DB_USER DB_PASS HUSHLINE_DEBUG_OPTS SQLALCHEMY_DATABASE_URI REGISTRATION_CODES_REQUIRED FLASK_ENV
 
 .PHONY: help
-help: ## Print the help message
+help:
 	@awk 'BEGIN {FS = ":.*?## "} /^[0-9a-zA-Z_-]+:.*?## / {printf "\033[36m%s\033[0m : %s\n", $$1, $$2}' $(MAKEFILE_LIST) | \
 		sort | \
 		column -s ':' -t
 
+.PHONY: install-deps
```
Comment (brassy-endomorph) [inline Makefile:22]: This one is def an ideological nit, but I try to keep the "repeatable scripts" of a `Makefile` to the ones that are tedious to type or vary across projects. What is the exact set of lints and tests the project should run? This project vs. another? But in every project with `poetry`, the command `poetry install` is going to be the same for the dev env. So I guess the rule I use is "things that are too long to type" or "things that aren't obvious."

**Review by brassy-endomorph — CHANGES_REQUESTED** [review]
Some changes are needed, and some are just nits

File: Makefile [hunk 2]
```diff
@@ -1,32 +1,58 @@
 .DEFAULT_GOAL := help
 FLASK_APP = hushline/__init__.py
+REDIS_URI ?= memory://
```
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply Makefile]: Updated with latest commit

File: Makefile [hunk 3]
```diff
@@ -1,32 +1,58 @@
...
+DB_PASS ?= dbpassword
+HUSHLINE_DEBUG_OPTS ?= 0
+SQLALCHEMY_DATABASE_URI ?= sqlite:///$(CURDIR)/hushline.db
+REGISTRATION_CODES_REQUIRED ?= False
+FLASK_ENV ?= development
+
+export FLASK_APP REDIS_URI SECRET_KEY ENCRYPTION_KEY DB_NAME DB_USER DB_PASS HUSHLINE_DEBUG_OPTS SQLALCHEMY_DATABASE_URI REGISTRATION_CODES_REQUIRED FLASK_ENV
 
 .PHONY: help
-help: ## Print the help message
```
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply Makefile:5]: Added comments back

File: Makefile [hunk 4]
```diff
@@ -1,32 +1,58 @@
...
+		poetry run black --check . && \
+		poetry run flake8 . && \
+		poetry run mypy .
 
 .PHONY: fmt
-fmt: ## Format the code
-	isort . && \
-		black .
+fmt:
+	poetry run isort . && \
```
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply Makefile]: updated in latest commit

File: hushline/ext.py [hunk 5]
```diff
@@ -1,11 +1,16 @@
+import os
+
 from flask_bcrypt import Bcrypt
 from flask_limiter import Limiter
 from flask_limiter.util import get_remote_address
 
 bcrypt = Bcrypt()
 
+# Check for an environment variable or use in-memory storage as fallback
+REDIS_URI = os.getenv("REDIS_URI")
```
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply hushline/ext.py]: Updated in last commit

File: install.sh [hunk 6]
```diff
@@ -138,83 +138,119 @@ export PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring
...
+        1)
+            echo "Setting REDIS_URI for development (in-memory storage)..."
+            REDIS_URI="memory://"
+            ;;
+        2)
+            REDIS_SERVER_URI=$(whiptail --inputbox "Enter your Redis server URI" 8 78 "redis://localhost:6379" --title "Redis Server URI" 3>&1 1>&2 2>&3)
+            echo "Setting REDIS_URI for production..."
+            REDIS_URI=$REDIS_SERVER_URI
+            ;;
+        *)
```
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply install.sh:224]: Since we're moving away from the install file I'll resolve this convo.

File: Makefile [hunk 7]
```diff
@@ -1,32 +1,58 @@
...
 
+.PHONY: install-deps
+install-deps:
+	poetry install
+
 .PHONY: run
-run: ## Run the app
-	FLASK_APP=$(FLASK_APP) flask run --debug -h localhost -p 5000
+run:
+	@export SECRET_KEY=$(SECRET_KEY) ; \
```
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply Makefile]: Completed with earlier commit.

File: Makefile (hunk 1 above)
Comment (glenn-sorrentino) [reply Makefile:7]: 👍 Removed from the makefile. We're also explicit about `poetry install` in the docs, so covered.

File: hushline/routes.py [hunk 8]
```diff
@@ -232,8 +228,8 @@ def register() -> Response | str:
                 flash("💔 Username already taken.", "error")
                 return redirect(url_for("register"))
 
-            password_hash = bcrypt.generate_password_hash(password).decode("utf-8")
-            new_user = User(primary_username=username, password_hash=password_hash)
+            new_user = User(primary_username=username)
+            new_user.password_hash = password
```
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply hushline/routes.py:232]: Updated with latest commit.

File: migrations/versions/577b927c0f49_.py
[the remark this answers is not in the document]
Comment (glenn-sorrentino) [reply migrations/versions/577b927c0f49_.py:1]: Outdated with rebase.

File: Makefile [hunk 9]
```diff
@@ -1,6 +1,3 @@
-.DEFAULT_GOAL := help
```
Comment (brassy-endomorph) [inline Makefile:1]: `.DEFAULT_GOAL` should stay. `help` is the first target, so `make` defaults to using it, but the presence of this serves as an explicit callout that this is the intention.

File: env.sh [hunk 10]
```diff
@@ -0,0 +1,11 @@
+export FLASK_APP="hushline/__init__.py"
+export REDIS_URI="memory://"
+export SECRET_KEY=$(openssl rand -hex 32)
+export ENCRYPTION_KEY=$(python3 -c 'from cryptography.fernet import Fernet; print(Fernet.generate_key().decode())')
```
Comment (brassy-endomorph) [inline env.sh:4]: We should just hardcode secrets here so that logins don't break every time the app restarts. If the GH action that checks for secrets yells at us, we shod make an exception.

File: Makefile [hunk 11]
```diff
@@ -24,8 +22,8 @@ fmt: ## Format the code
 		black .
 
 .PHONY: init-db
-init-db: ## Initialize the dev database
-	FLASK_APP=$(FLASK_APP) flask db-extras init-db
+init-db:  ## Initialize the dev database
```
Comment (brassy-endomorph) [inline Makefile:25]: Nit: unnecessary whitespace change

File: migrations/versions/a74facb1f029_.py
Comment (brassy-endomorph) [inline migrations/versions/a74facb1f029_.py:1]: This migration is going to need to be reworked and shouldn't be part of this PR. The `flask db-extras init-db` command doesn't rely on `alembic` so we can just leave this out for now

File: Makefile [hunk 12]
```diff
@@ -9,7 +6,8 @@ help: ## Print the help message
 
 .PHONY: run
 run: ## Run the app
-	FLASK_APP=$(FLASK_APP) flask run --debug -h localhost -p 5000
+	@source ./env.sh ; \
```
Comment (brassy-endomorph) [inline Makefile:9]: The semicolon should be `&&` so that if the `source` fails the target fails.

**Review by brassy-endomorph — CHANGES_REQUESTED** [review]

File: Makefile (hunk 9 above)
Comment (glenn-sorrentino) [reply Makefile:1]: ✅

File: Makefile (hunk 11 above)
Comment (glenn-sorrentino) [reply Makefile:25]: ✅

File: Makefile (hunk 12 above)
Comment (glenn-sorrentino) [reply Makefile]: ✅

File: migrations/versions/a74facb1f029_.py
Comment (glenn-sorrentino) [reply migrations/versions/a74facb1f029_.py:1]: ✅

File: docs/DEV.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+### Mac
+
+| Tested Platform | Date |
+|-|-|
+| Mac M2 | Apr. 2024 |
+
+#### Install Packages
+- `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"`
+- `brew install python git git-lfs nginx redis rust`
```
Comment (brassy-endomorph) [inline docs/DEV.md:19]: We shouldn't need `nginx` for dev

File: docs/DEV.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+- `echo 'export PATH="/opt/homebrew/bin:$PATH"' >> ~/.zshrc`
+- `curl -sSL https://install.python-poetry.org | python3 -`
```
Comment (brassy-endomorph) [inline docs/DEV.md:21]: `brew install poetry` might make more sense, no?

File: docs/DEV.md [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+- `source ~/.zshrc`
+
+#### Clone the Repo
+- `https://github.com/scidsg/hushline.git`
+- `cd hushline`
+- `git switch dev-env` # Temporary until we merge back into Main
+- `python -m venv venv`
+- `source venv/bin/activate`
+- `poetry install`
+- `source env.sh`
```
Comment (brassy-endomorph) [inline docs/DEV.md:32]: These last lines should just be `poetry run make init-db run`. We shouldn't run the migrations in dev as part of the main workflow.

File: migrations/README
Comment (brassy-endomorph) [inline migrations/README:1]: You can yank the commit where you deleted the `migrations` dir. They were already there. Just the commit where you added the initial migration probably should go

File: docs/DEV.md (hunk 13 above)
Comment (glenn-sorrentino) [reply docs/DEV.md:19]: Right. Oversight.

File: docs/DEV.md (hunk 14 above)
Comment (glenn-sorrentino) [reply docs/DEV.md:21]: 👍

File: docs/DEV.md (hunk 15 above)
Comment (glenn-sorrentino) [reply docs/DEV.md]: ✅

File: migrations/README
Comment (glenn-sorrentino) [reply migrations/README:1]: Reverted the last commit where we deleted migrations, and the initial migration commit.

File: docs/DEV.md (hunk 15 above)
Comment (glenn-sorrentino) [reply docs/DEV.md]: ✅

**Comment by glenn-sorrentino** [comment]
@brassy-endomorph I cleaned this PR up a bunch. It should be easier to review now.
